### PR TITLE
test: transfer to system actor

### PR DESCRIPTION
**Pull request by alanshaw**
~~Currently failing, should this be allowed?~~

Update: this currently is allowed (as the tests prove), but is being considered to be not allowed: https://github.com/filecoin-project/specs/issues/1069

refs #13

Depends on:

* [x] #71

### CODE REVIEW

**Comment by raulk** [comment]
Discussed in colo today. Since these transfers are currently allowed, let's change the asserts to `exitcode.Ok` and lock in these test vectors. If the behaviour changes, these vectors will break and that is totally legit.

**Comment by alanshaw** [comment]
Currently getting the following 2 failures, reward and burnt funds actors receiving a mysterious `1_000_000_000` and `100_000_000_000` extra FIL recpectively:

Screenshot 2020-08-24 at 15 04 55

**Comment by raulk** [comment]
@alanshaw that's actually correct! The gas basefee is burnt (that’s why you see an increment in the BurntFundsActor), and the miner rewards are buffered in the RewardActor, where the accumulate throughout the processing of a block. When the block is processed, they are transferred to the miner's account.

File: gen/suites/transfer/main.go [hunk 1]
```diff
@@ -155,4 +157,67 @@ func main() {
...
+		},
+		&MessageVectorGenItem{
+			Metadata: &Metadata{
+				ID:      "to-reward-actor",
+				Version: "v1",
+			},
+			Func:  transferToSystemActor(builtin.RewardActorAddr),
+			Mode:  ModeLenientAssertions,
+			Hints: []string{schema.HintIncorrect},
+		},
```
Comment (raulk) [inline gen/suites/transfer/main.go:184]: Gotta sort this out as we should be asserting the actual balance taking into account mining rewards.

File: gen/suites/transfer/main.go [hunk 2]
```diff
@@ -155,4 +157,67 @@ func main() {
...
+		&MessageVectorGenItem{
+			Metadata: &Metadata{
+				ID:      "to-burnt-funds-actor",
+				Version: "v1",
+			},
+			Func:  transferToSystemActor(builtin.BurntFundsActorAddr),
+			Mode:  ModeLenientAssertions,
+			Hints: []string{schema.HintIncorrect},
+		},
```
Comment (raulk) [inline gen/suites/transfer/main.go:221]: Gotta sort this out as we should be asserting the actual balance taking into account gas burnt during message processing.

File: gen/suites/transfer/main.go [hunk 3]
```diff
@@ -155,4 +159,33 @@ func main() {
...
+		{name: "verified-registry", addr: builtin.VerifiedRegistryActorAddr},
+		{name: "burnt-funds", addr: builtin.BurntFundsActorAddr, extraFunc: func(r *vm.ApplyRet) big.Int {
+			return CalculateBurn(gasLimit, r.GasUsed)
+		}},
+	}
+
+	var sysReceiverItems []*MessageVectorGenItem
+	for _, a := range sysActors {
+		sysReceiverItems = append(sysReceiverItems, &MessageVectorGenItem{
+			Metadata: &Metadata{ID: "to-" + a.name + "-actor", Version: "v1"},
```
Comment (raulk) [inline gen/suites/transfer/main.go:185]: Super nit: I think `mt.Sprintf("to-%s-actor", a.name)` would read better here.

File: gen/suites/transfer/system_receiver.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+// to a system actor. The (optional) calcExtra parameter calculates an amount to
+// expect to see in the receiver's account due to the fact that the receiver is
+// a special actor that is sent funds during the process of a transfer.
+//
+// e.g. the reward actor receives the miner tip (gas limit * gas premium), which
+// is held for the block miner.
+//
+// e.g. The burnt funds actor receives the gas burn base fee.
+//
+// TODO: These tests may break in the future if sending to a system actor
```
Comment (raulk) [inline gen/suites/transfer/system_receiver.go:24]: Nit: can we add this as a comment on the test vector metadata?

**Review by raulk — APPROVED** [review]

**Comment by raulk** [comment]
Boat-putting is advised 🚢

**Pull request merged**